### ISSUE TITLE
[FEATURE] Better handling of DB connection issues

### ISSUE THREAD

**Issue by kevinpapst** [tobybatch/kimai2#174]
**Is your feature request related to a problem? Please describe.**

Actually while hunting the issue in #172 I was thinking:
- this loop has no break condition if it doesn't respond for eg 20 times
- this loop isn't very helpful in finding a possible cause: to me (without testing) it looks as even a typo in the password could lead to an infinite loop?

**Describe the solution you'd like**

I wrote a little php script, which could be used in the `until` loop in https://github.com/tobybatch/kimai2/blob/main/startup.sh#L24 instead of the plain PDO call.

In case that the connection could not be established, it shows an error message and returns an exit code.
Exit codes `4` and `5` could be used to retry it X times (X = a number we should define), the other ones could IMO directly lead to an exit of the startup script.

```php
<?php
try {
    $pdo = new \PDO('mysql:host=127.0.0.1;dbname=kimai2;port=3306', 'root', 'password', [
        \PDO::ATTR_ERRMODE => \PDO::ERRMODE_EXCEPTION
    ]);
} catch(\Exception $ex) {
    switch ($ex->getCode()) {
        // we can immediately stop startup here and show the error message
        case 1045:
            echo 'Access denied';
            die(1);
        // we can immediately stop startup here and show the error message
        case 1049:
            echo 'Unknown database';
            die(2);
        // a lot of errors share the same meaningless error code zero
        case 0:
            // this error includes the database name, so we can only search for the static part of the error message
            if (stripos($ex->getMessage(), 'SQLSTATE[HY000] [1049] Unknown database') !== false) {
                echo 'Unknown database';
                die(3);
            }
            switch ($ex->getMessage()) {
                // eg. no response (fw) - the startup script should retry it a couple of times
                case 'SQLSTATE[HY000] [2002] Operation timed out':
                    die(4);
                // special case "localhost" with a stopped db server (should not happen in docker compose setup)
                case 'SQLSTATE[HY000] [2002] No such file or directory':
                    echo 'Connection could not be established';
                    die(5);
                // using IP with stopped db server - the startup script should retry it a couple of times
                case 'SQLSTATE[HY000] [2002] Connection refused':
                    echo 'Connection refused';
                    die(5);
            }
            echo $ex->getMessage();
            die(7);
        default:
            // unknown error
            echo $ex->getMessage();
            die(10);
    }
}
```

What do you think @tobybatch @Schrolli91 ?

Edit: please note that I will not add it myself, as I have no time to learn how to use/work with docker compose right now. This is just an idea for a possible solution 😄

**Comment by Schrolli91** [comment]
Sounds good to me. 
We realy should implement something like this to get better feedback if there some problems with the DB on startup.

**Comment by tobybatch** [comment]
agreed, I'll look into it.

**Comment by tobybatch** [comment]
closed by #213
